### DISCUSSION TITLE
If there was one thing you could change about LuckPerms, what would it be?

### DISCUSSION THREAD

**Discussion by lucko** [LuckPerms/LuckPerms#2824] [General]
Hello! :wave:

Welcome to GitHub Discussions - isn't this fancy!

I thought we could try and use this new discussion functionality to fill the gap between conversations on Discord (which tend to get left behind and lost quite quickly!) and the regular ol' issues tab, which is strictly for bug reports and feature suggestions.

So, to get things started, I have a question for you!

___

### If there was one thing you could change about LuckPerms, what would it be?

I'm thinking things along the lines of:

* is there something you **frequently** do with LuckPerms that could be **streamlined** in some way? ⏩
* are there any **little annoyances** you've experienced with the plugin that could be improved?! 🔨
* is there anything simple that we can change to make using the plugin that little bit **easier/better for you** to use? 💡
* is there a feature you just absolutely **hate and wish we had never added** - let us know!! 💣

.. but really, these are just some ideas, anything is valid! 

___

Please react with :+1: ​ to ideas you agree with, and :-1: ​ to ideas you don't!

Feel free to reply to comments if you have something to add. Looking forward to seeing what you guys have to say! <3

**Comment by Turbotailz** [comment]
I'd change the name to LuckyPerms

**Reply by tadhgboyle** [reply]
Thank you for being courageous and finally speaking out about this issue.

**Reply by mcmdev** [reply]
I know way to many people who always call it LuckyPerms. Might be the easiest solution

**Reply by Handkep** [reply]
Absolutely!

**Comment by TrollyLoki** [comment]
The permission check command only looks at specifically defined permissions, it would be helpful to have a version of it that factors in things like wildcard permissions.

**Reply by emilyy-dev** [reply]
<s>The [`lp search`](https://luckperms.net/wiki/General-Commands#lp-search-comparison-permission) command can take an (optional) "comparison mode" argument exactly like the [`lp bulkupdate`](https://luckperms.net/wiki/Bulk-Editing#constraints) command does for constraints, so you can narrow your search down with wildcards (`_` for single char wildcard, `%` for multi chars wildcard) :slightly_smiling_face: Is that more-so what you were looking for?</s>


Wow never mind I lack the ability of reading apparently.

**Reply by lucko** [reply]
You make a good point!

The systems behind permission checking have actually improved quite a lot since the various `check` commands were written. These improvements are visible in the Verbose viewer but were never integrated into the permission check command..

So yes, I agree! Definitely an area that could be made better quite easily - thank you for the idea!

**Reply by lucko** [reply]
I've had a go at implementing this!

![Screenshot 2021-01-13 at 16 15 50](https://user-images.githubusercontent.com/8352868/104482140-7fba6000-55be-11eb-96e5-7bb48d0dae26.png)

So, what are we looking at here.. well:

* The first section (before the blank line) shows information about the places the user could have gotten the permission from, either if it has been set directly, or inherited
  * This consolidates the output of the previous `/lp user <u> permission check` and `/lp user <u> permission checkinherits` commands
  * It is ordered by priority! Entries higher up in the list will take priority over those lower down.
* The second section shows the result of a permission check, as well as the processor that ended up responding to the check.
  * The "Result" bit was previously available from the `/lp check` command, but the other sections are new.
  * Processors are what implement things like wildcards, regex, defaults etc on top of standard permissions. `MapProcessor` just means a direct match was found in the users permissions map.

All of this is now under one command instead of three, `/lp user <user> permission check`. (and it works for groups too!)

As for what the output shows.. well, we can see that there has been a misconfiguration! "luck", an admin, inherits the true version of the permission, but this is actually being overridden by the false setting in the default group. So, we probably need to setup some group weights!

![Screenshot 2021-01-13 at 16 17 38](https://user-images.githubusercontent.com/8352868/104482131-7f21c980-55be-11eb-9d97-c2f6c5d9b1ef.png)

The second screenshot shows some of the other processors in action - firstly the Bukkit defaults processor, then, after setting a wildcard for `bukkit.command.*`, the result of the WildcardProcessor working.

I think definitely an improvement over what we had before!

**Reply by tadhgboyle** [reply]
Looks awesome!

**Reply by lucko** [reply]
Also now includes relevant wildcard permissions!

![Screenshot 2021-01-13 at 17 28 48](https://user-images.githubusercontent.com/8352868/104487321-dc207e00-55c4-11eb-802c-2fd2c84ba6e8.png)

**Reply by TrollyLoki** [reply]
Awesome! This is exactly the kind of thing I was looking for and more!

**Reply by ThatBubbledGuy** [reply]
Wow cool!

**Comment by ItsCryptic** [comment]
LuckPerms telling you it has an update, at the moment there isn't a system in place and you are pretty much reliant on reading #changelog for whenever an update comes. I feel like either some kind of message in console when the server boots saying you're using an outdated version or something could be nice.

**Reply by Turbotailz** [reply]
Considering that updates are very frequent, this could get a bit annoying for operators who don't tend to update things very often.

Would be a good config option though! I was also going to add a little notification to the editor at some point, too.

**Reply by ItsCryptic** [reply]
Yeah it would get annoying but it would be nice for the people who care to keep there plugins up to date or lp having an extension for auto updates that would just get installed whenever the server reboots.

**Reply by Wqrld** [reply]
im always extremely irritated by these update logs. Either you just dont see them anymore or they clog up your console and chat.

**Reply by broken1arrow** [reply]
I think is good idea, I always at least 40+ versions behind because of this. 

But in config add setting "checkforupdate: "true or false"" . So if someone not like it you never see it.

I even goo soo fare to say a ingame notis, is even better. But with only of you has luckperms.updatenotis all other plugin some has it ingame has a perm like this (I know some has * perm in lucky, but set it to false shall work if you not like it or you has the perm outside of normal luck.* perms, so you have to add this extra perm for see ingame notis).

**Reply by Virologicbody** [reply]
100% agree with this, but with the option to disable in config.. so works for those who are for and against.
I personally like to know when there are plugin updates, but also agree that they can be annoying.. if anyone knows how often CMI updates!

**Reply by ThatBubbledGuy** [reply]
This is codes that used to make minecraft!?

**Comment by isteiger** [comment]
I want the prefixes to appear above a players' head.

**Reply by Turbotailz** [reply]
You can do that... With a name tag plugin. It isn't the job of LuckPerms to display prefixes anywhere.

**Reply by isteiger** [reply]
Yeah, good point...

**Comment by tristan-f-r** [comment]
Velocity-global permissions. A way for different instances to communicate with each other using luckperms and hook into velocity globally. I use multiple different platforms for my network, and having that would be a lifesaver.

**Reply by tadhgboyle** [reply]
I assume by "different instances" you mean Bukkit, Sponge etc?

**Reply by tristan-f-r** [reply]
Mhm

**Comment by ZomkaDEV** [comment]
I would've renamed it into FloofPewms (uwu)

**Reply by underscore11code** [reply]
-_-

**Comment by Wqrld** [comment]
Make the commands a bit easier (aliases?) for newcomers and market more on the /lp editor feature.

I love luckperms but i have seen a lot of people struggling because they are not what you expect from other plugins. E.g. alias parent to group and make it more clear on the plugin page how easy the web editor is.

**Reply by Wqrld** [reply]
also: allow `/lp user wqrld permission add bukkit.*` instead of only `permission set.`

**Reply by tadhgboyle** [reply]
I have a feeling adding more ways to do the same task will only confuse people more

**Reply by Wqrld** [reply]
full-on renaming would be worse for existing implementations so i feel like this would be the best "inbetween" option. But for the people ive had to help this was one of the major issues coming from pex/groupmanager

**Reply by benwoo1110** [reply]
> also: allow `/lp user wqrld permission add bukkit.*` instead of only `permission set.`

This will only make it more confusing with 2 different argument doing the same thing. Basing off your example, there is `/lp user USER parent add` vs `/lp user USER parent set` which has differing behaviour.

Also, this will get increase the amount to documentation needed for the added alias.

**Reply by Turbotailz** [reply]
There used to be these aliases you spoke of in version 4 but they were removed in v5 with the following reason:

> **A number of "undocumented" command aliases have been removed.**
> These were confusing to new users, and made it easy to mistakenly run a command that did x when you wanted y.

I'd also like to point out how ridiculous this command would be:

```/lp group <group> group add <group>```

:P

**Comment by hellot1m** [comment]
I'd suggest a description database similar to ultra permissions that the community can update together. Then allowing a hover feature or some kind of feature to utilize this so that users don't have to go back and forth for perm descriptions. 

Also maybe a way to list all available perms from a plugin in the web UI. Having to type out each one, especially if there are a lot of perms in a plugin can get extremely tiresome.

**Reply by broken1arrow** [reply]
the system they use to day, use some type of cache (when plugin sends perm checks). 

One thing should bee better is tabcomplete in editor, some not add the perm to the list (so you can edit it before you add it).

**Reply by Turbotailz** [reply]
Personally I find it should be the job of the plugin developers to document their permissions properly. It shouldn't be LuckPerms' job to tell users what permissions are meant to do. Not to mentioned the absurd amount of work required to get this to function properly for most popular plugins.

Also as @broken1arrow mentioned about permissions being cached - this means that your suggestion of the Editor showing only certain permissions from a plugin is not really going to work when the suggested permissions on the Editor is almost never a complete list. I agree that it can be hard to find the right permissions but maybe you should just try `/lp tree` instead?

**Comment by hxpd** [comment]
I would implement a Tablistprefix System.

**Reply by Turbotailz** [reply]
Unnecessary, when there are plugins that do this already and it's not the job of LP to display prefixes/suffixes anywhere.

**Comment by TKDKid1000** [comment]
Can't think of anything honestly! This plugin is already all I could ask for and more!

**Comment by Jerimu** [comment]
The possibility to give people permission to check certain groups instead off all groups. Like all groups that start with survival_ could be inspected with certain permissions.

**Reply by Turbotailz** [reply]
This is likely already possible with a combination of [Argument Based Command Permissions](https://luckperms.net/wiki/Argument-based-command-permissions) and [Regex permissions](https://luckperms.net/wiki/Advanced-Setup#regex).

**Comment by tadhgboyle** [comment]
The ability to export `/lp logs` to a flat file (json, csv, etc) with an optional param for how many y/m/d/m/s to go back, or a unix timestamp to start at.

**Reply by Turbotailz** [reply]
Maybe even an online viewer 😉

**Reply by tadhgboyle** [reply]
Ohhh yeah￼, nice idea.
Could it function similar to git too? Like with undo/redo functions? I can see the complexity increase significantly in the development side, but it would probably be very used. I'd be happy to talk about this in discord to brainstorm ui/ux

**Comment by Nutmegan** [comment]
I'm a little new to running servers, so this idea may not be smart, or maybe there is something I'm missing. I wish there were a way in LP to force nasty commands that do not have permissions to behave better. (I'm specifically thinking of WorldEdit, but there are a few others.) It there were a way to assign the command a permission and then allow LP to manage the custom permission, it would simplify my current process. Right now, I use a combination of command.yml (if I don't want it used at all) or PL-Hide to manage the permissions of the group that can use it. I tried using MyCommand, but I never could get it to work just right.

As a next step, I would love a feature to alias commands and hide them and skip command.yml. Really, that's not LPs job, but it would make my life simpler being able to manage all that in one place.

Again, LP has a lot of functions I know I'm not totally tapping into yet because it was too much to take in all at once. Maybe this can already be done and if so, yay! I would love to hear more about how you would go about it.

edited typos

**Comment by martijnpu** [comment]
Add support for the pgup/pgdn (page-up/page-down) buttons to scroll in the web-editor.

I'm someone who barly uses the mouse and try do as much as possible with the keyboard, mostly because the keyboard is a lot faster if you know what you're doing. But I just came to the conculsion that the web-editor doesn't support the pgup/pgdn buttons to scroll very fast trough the long list of permissions. I know the arrow keys are used to select one above or beneath, but I want something faster.

**Reply by Turbotailz** [reply]
> to scroll very fast trough the long list of permission

Which list do you mean, exactly? The group's/user's list of permissions, or the list in the dropdown menu when you add perms?

You can use pg-up/down in the main window but you have to focus it first by clicking on a node. The "add node" permission list can't be focused due to the nature of the input component that's being used - I need to replace that with something a bit more usable but can't say when that will happen. 

More info on how you'd like to use the keyboard to navigate would be pretty good 👍 (I'd also recommend opening an issue on the lucko/LuckPermsWeb repo on the subject)

**Reply by martijnpu** [reply]
I mean the groups/users list of permission. 
The dropdown is irrelevant because if you don't know you need to read them all, or if you know you type it.

> You can use pg-up/down in the main window but you have to focus it first by clicking on a node.
That's true, but it unfocussed when it's off your screen and the pgup/pgdn will stop working.

Besides that maybe the ability like scrolling with your mouse; when you hover the box you can control it with pgup/pgdn.

**Reply by Turbotailz** [reply]
> That's true, but it unfocussed when it's off your screen and the pgup/pgdn will stop working.

Hmm, yeah unfortunately that's a side-effect of the solution I used to minimise lag for users with thousands of nodes in a single group/user. Possibly I could fix it by putting space around the nodes where you can click without focusing on a certain node but still focus the list.

> Besides that maybe the ability like scrolling with your mouse; when you hover the box you can control it with pgup/pgdn.

Sadly I don't think that's possible with how browser focus works. I'm sure I can figure something out though.

**Comment by TKDKid1000** [comment]
Maybe the ability to add multiple permission nodes at once (in a single command). Kind of like worldedit does it where you separate them with commas. So like `/lp group default permission set essentials.kit,essentials.kits.pvp,essentials.warp,essentials.warps.main`

**Reply by martijnpu** [reply]
Why don't you use the web-editor? There you can add multiple nodes at the same time.

**Reply by TKDKid1000** [reply]
> Why don't you use the web-editor? There you can add multiple nodes at the same time.

Well, it's easier and quicker to just do a command. Instead of opening a new editor session every time. I only really use the editor when I'm doing large changes (changing every groups prefix for example)

**Reply by Turbotailz** [reply]
I think this is a very valid suggestion. It could even work with creating groups, adding parents, etc.

**Comment by TKDKid1000** [comment]
A permission node, that makes it so you can only add certain nodes/groups. Say you want group admin to be able to promote other people to any rank including admin, but not owner. You can’t do that. With this it could be some kind of permission like `luckperms.permissions.group.admin` So you can only set someone group to admin. Just an idea.

**Reply by Turbotailz** [reply]
[You actually can do that.](https://luckperms.net/wiki/Argument-based-command-permissions)

**Comment by MikalusPlayz** [comment]
A way to toggle permission nodes for a specific group, depending on if another group has members online or not.

For example: Builders don't get worldedit permission, only Creative. But if a Higher-Up is online, it toggles them to be able to use WorldEdit. And if no Higher-Ups are one, the permissions to use WorldEdit get disabled.

**Reply by TKDKid1000** [reply]
> A way to toggle permission nodes for a specific group, depending on if another group has members online or not.
> 
> For example: Builders don't get worldedit permission, only Creative. But if a Higher-Up is online, it toggles them to be able to use WorldEdit. And if no Higher-Ups are one, the permissions to use WorldEdit get disabled.

Definitely would use this. Didn't think of it but it's awesome! Could be as simple as another context like `is-online=(user)` and `is-online-group=(group)`. And with this maybe `is-offline=(user)` and `is-offline-group=(group)`

**Reply by DarkKingKaan** [reply]
Totally agree! I would absolutely love to use this kind of feature!

**Comment by TKDKid1000** [comment]
I've recently been needing this feature and it would be really handy, the ability to check for false contexts. I wanted this one user to have a certain group, when in all but one server. To do this I had to manually add each check for every server EXCEPT the one that I wanted them not to have it. So I think it would be good if we could have contexts like `world=!worldname` and `server=!servername`. Just thought I'd ask.

**Comment by Jerimu** [comment]
Master, Slave thing. So you can block luckperms commands from console (except bulkupdate ofc) so you could have only specific servers run luckperms commands from console.

**Comment by ThatBubbledGuy** [comment]
GitHub disscussions are very good  for people!

**Comment by DarkKingKaan** [comment]
The ability to specify which group to promote or demote a player on a certain track.

Currently, the command for promoting/demoting goes like `/lp user <user> promote/demote <track>`, and it simply increases or decreases the player's group by one, as it should.

I would like the command to have an additional, optional variable: `/lp user <user> promote/demote <track> <group>`, which would allow us to skip a few groups on a track, so we don't have to spam the command multiple times when trying to promote someone to a group in a track that has other groups before it.

This would also be useful to prevent accidental demotions/promotions of players. If you try to *promote* a player to a group that is lower than his current group, the command wouldn't work. If you try to *demote* a player to a higher group than his current group, the command also wouldn't work.

A perfect example of this is, sadly outdated, GroupManager. I have used it for a long time before finally switching to LuckPerms and I haven't looked back since. But this one little feature is missing. And GM had it. 

This is my suggestion for the plugin. Feel free to leave any feedback or comments.

**Reply by Turbotailz** [reply]
I'm sorry that you had to write all of that only to find out that the [parent settrack](https://luckperms.net/wiki/Parent-Commands#lp-usergroup-usergroup-parent-settrack-track-indexgroup-context) command does exactly what you are asking for 😅

**Reply by DarkKingKaan** [reply]
Not quite, the `settrack` command can still accidentally demote you to a lower rank. If, for example, I would have a crate setup where you can get a rank, a VIP+ could open the crate, win a "VIP" rank and get demoted. As I stated, the feature would work like this: If you try to promote a player to a group that is lower than his current group, the command wouldn't work. If you try to demote a player to a higher group than his current group, the command also wouldn't work.

**Comment by ReturnDesender** [comment]
I would suggest the ability to set a required amount of playtime in order to be promoted to a specific group. I have a small server with friends, but we are open to inviting friends of friends etc. I don't want people to be able to join and immediately bypass spawn protections or anything like that. I want to turn on spawn protection but not have to give people who join for 5 minutes access to possibly grief us, and I am not home enough to manually add additional people.

**Comment by ReturnDesender** [comment]
I am running an economy on my server as well, and I was wanting to try to sell permissions such as additional sethomes and things of that sort to players. The only way I could think to do this since there does not seem to be any option for something like this was to use command blocks. It does not seem that command blocks have access to use LuckPerms.

**Comment by Turbotailz** [comment]
@ReturnDesender these are great ideas but they are IMO out of scope for a permission plugin. There are other plugins available which can help accomplish these things.

**Comment by Seeker1437** [comment]
Allow submitting multiple contexts like I can with permissions.

I sadly have a high number of contexts that I use, its sometimes dreadful to know if I need to add a new pernission node I sometimes have to click through like 12 contexts to add them.

Would be noice to be able to paste in a context using a pair format.

`context=value` or `context:value`
